### PR TITLE
upload: add before-remove function

### PR DESCRIPTION
**Pull request by firesh**
upload: add before-remove function. It can support a function returns a boolean or Promise.
#8399 requests this feature.

Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide ([中文](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.zh-CN.md) | [English](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.en-US.md) | [Español](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.es.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer relative issues for you PR.

### CODE REVIEW

**Comment by firesh** [comment]
@Leopoldthecoder Both Chinese and English document is update. But I was not able to translate it to Spanish. So I just Keep it for others folks.

File: packages/upload/src/index.vue [hunk 1]
```diff
@@ -175,10 +176,30 @@ export default {
...
+        return;
+      }
+
+      const before = this.beforeRemove(file, this.uploadFiles);
+      if (before && before.then) {
+        before.then(() => {
+          doRemove();
+        }, () => {
+          // do nothing
+        });
```
Comment (Leopoldthecoder) [inline packages/upload/src/index.vue]: You can use `noop` here:
```JS
before.then(() => {
  doRemove();
}, noop);
```

File: packages/upload/src/index.vue [hunk 2]
```diff
@@ -175,10 +176,30 @@ export default {
...
+      } else if (before !== false) {
+        doRemove();
+      } else {
+        // do nothing
+      }
```
Comment (Leopoldthecoder) [inline packages/upload/src/index.vue]: This `else` block can be removed entirely.

File: examples/docs/en-US/upload.md [hunk 3]
```diff
@@ -116,6 +116,15 @@
       },
       handleExceed(files, fileList) {
         this.$message.warning(`You can upload up to 3 files. You selected ${files.length} files this time, and ${files.length + fileList.length} files totally`);
+      },
+      handleBeforeRemove(file, fileList) {
+        return new Promise((resolve, reject) => {
+          if (confirm('It looks good, are you sure to remove?')) {
+            resolve()
```
Comment (Leopoldthecoder) [inline examples/docs/en-US/upload.md]: Please add a semicolon here. Also applies to the other 7 `resolve()`s and `reject()`s.

**Comment by Leopoldthecoder** [comment]
Thanks for contributing.

**Pull request merged**